datasource/ethereum/src/block_stream.rs

```diff
@@ -1096,5 +1204,188 @@ fn create_log_filter_from_subgraph(manifest: &SubgraphManifest) -> EthereumLogFi
+    }
+}
+
+fn create_block_filter_from_subgraph(manifest: &SubgraphManifest) -> Option<EthereumBlockFilter> {
+    let block_filter = manifest
+        .data_sources
+        .iter()
+        .filter(|data_source| {
+            data_source
+                .mapping
+                .block_handler
+                .is_some() && data_source.
+                source.
+                address.
+                is_some()
+        })
+        .map(|data_source| {
+            data_source.source.address.unwrap()
+        })
+        .collect::<EthereumBlockFilter>();
+    match block_filter.contract_addresses.len() {
+        0 => return None,
+        _ => return Some(block_filter),
+    }
+}
+
+fn parse_triggers(
+    log_filter_opt: Option<EthereumLogFilter>,
+    call_filter_opt: Option<EthereumCallFilter>,
+    block_filter_opt: Option<EthereumBlockFilter>,
+    descendant_block: &EthereumBlockWithCalls,
+) -> Result<Vec<EthereumTrigger>, Error> {
+    let mut triggers = Vec::new();
+    triggers.append(&mut parse_log_triggers(
+        log_filter_opt,
+        &descendant_block.ethereum_block,
+    ));
+    triggers.append(&mut parse_call_triggers(
+        call_filter_opt,
+        descendant_block,
+    ));
+    triggers.append(&mut parse_block_triggers(
+        block_filter_opt,
+        descendant_block,
+    ));
+    let tx_hash_indexes = descendant_block
+        .ethereum_block
+        .transaction_receipts
+        .iter()
+        .map(|receipt| {
+            (receipt.transaction_hash, receipt.transaction_index.as_u64())
+        })
+        .collect::<HashMap<H256, u64>>();
+
+    // Ensure all `Call` and `Log` triggers have a transaction index
+    for trigger in triggers.iter() {
+        match trigger {
+            EthereumTrigger::Log(log) => {
+                if !tx_hash_indexes.get(&log.transaction_hash.unwrap()).is_some() {
+                    return Err(format_err!("Unable to determine transaction index for Ethereum log."))
+                }
+            }
+            EthereumTrigger::Call(call) => {
+                if !tx_hash_indexes.get(&call.transaction_hash.unwrap()).is_some() {
+                    return Err(format_err!("Unable to determine transaction index for Ethereum call."))
+                }
+            }
+            EthereumTrigger::Block(_) => continue,
+        }
+    }
+
+    // Sort the triggers
+    triggers
+        .sort_by(|a, b| {
+            let a_tx_index = a.transaction_index(&tx_hash_indexes).unwrap();
+            let b_tx_index = b.transaction_index(&tx_hash_indexes).unwrap();
+            if a_tx_index.is_none() && b_tx_index.is_none() {
+                return Ordering::Equal
+            }
+            if a_tx_index.is_none() {
+                return Ordering::Less
+            }
+            if b_tx_index.is_none() {
+                return Ordering::Greater
+            }
+            a_tx_index.unwrap().cmp(&b_tx_index.unwrap())
+        });
```

Comment:
It looks like all calls within a given transaction will appear after the logs within that transaction. Is that desirable? Might be useful to document that behaviour since it's visible to users I think :)

Reply:
Good idea, I'll add some documentation in the code and the user docs

Reply:
Yeah, this is our initial convention until we can properly interleave these things.